components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2ScopeService.java

```diff
@@ -0,0 +1,373 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wso2.carbon.identity.oauth2;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.oauth.cache.OauthScopeCache;
+import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+import org.wso2.carbon.identity.oauth2.bean.Scope;
+import org.wso2.carbon.identity.oauth2.dao.ScopeMgtDAO;
+import org.wso2.carbon.identity.oauth2.util.Oauth2ScopeUtils;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * OAuth2ScopeService use for scope handling
+ */
+public class OAuth2ScopeService {
+    //TODO get the instance inside the constructor
+    private static ScopeMgtDAO scopeMgtDAO;
+    private OauthScopeCache scopeCache = OauthScopeCache.getInstance();
+    private static final Log log = LogFactory.getLog(OAuth2ScopeService.class);
+
+    public OAuth2ScopeService() {
+        scopeMgtDAO = ScopeMgtDAO.getInstance();
+    }
+
+    /**
+     * Register a scope with the bindings
+     *
+     * @param scope details of the scope to be registered
+     * @throws IdentityOAuth2ScopeException
+     */
+    public Scope registerScope(Scope scope) throws IdentityOAuth2ScopeException {
+        int tenantID = Oauth2ScopeUtils.getTenantID();
+
+        // check whether the scope name is provided
+        if (StringUtils.isBlank(scope.getName())) {
+            throw Oauth2ScopeUtils.handleClientException(Oauth2ScopeConstants.ErrorMessages.
+                    ERROR_CODE_BAD_REQUEST_SCOPE_NAME, null);
+        }
+
+        // check whether the scope description is provided
+        if (StringUtils.isBlank(scope.getDescription())) {
+            throw Oauth2ScopeUtils.handleClientException(Oauth2ScopeConstants.ErrorMessages.
+                    ERROR_CODE_BAD_REQUEST_SCOPE_DESCRIPTION, null);
+        }
+        try {
+            // check whether a scope exists with the provided scope name
+            String existingScopeID = scopeMgtDAO.getScopeIDByName(scope.getName(), tenantID);
+            if (existingScopeID != null) {
+                throw Oauth2ScopeUtils.handleClientException(Oauth2ScopeConstants.ErrorMessages.
+                        ERROR_CODE_CONFLICT_REQUEST_EXISTING_SCOPE, scope.getName());
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw Oauth2ScopeUtils.handleServerException(Oauth2ScopeConstants.ErrorMessages.
```

Comment:
Only implement handleServerException without data pram, append the data in message itself.